src/client/Microsoft.Identity.Client/AuthenticationInfoParameters.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http.Headers;
+using Microsoft.Identity.Client.Utils;
+
+namespace Microsoft.Identity.Client
+{
+    /// <summary>
+    /// Parameters returned by the Authentication-Info header. This allows for
+    /// scenarios such as proof-of-possession, etc.
+    /// See https://www.rfc-editor.org/rfc/rfc7615
+    /// </summary>
+    public class AuthenticationInfoParameters
+    {
+        private const string AuthenticationInfoKey = "Authentication-Info";
+        /// <summary>
+        /// The next nonce to be used in the preceding authentication request.
+        /// </summary>
+        public string NextNonce { get; private set; }
+
+        /// <summary>
+        /// Return the <see cref="RawParameters"/> of key <paramref name="key"/>.
+        /// </summary>
+        /// <param name="key">Name of the raw parameter to retrieve.</param>
+        /// <returns>The raw parameter if it exists,
+        /// or throws a <see cref="System.Collections.Generic.KeyNotFoundException"/> otherwise.
+        /// </returns>
+        public string this[string key]
+        {
+            get
+            {
+                return RawParameters[key];
+            }
+        }
+
+        /// <summary>
+        /// Dictionary of raw parameters in the Authentication-Info header (extracted from the Authentication-Info header
+        /// string value, without any processing). This allows support for APIs which are not mappable easily to the standard
+        /// or framework specific (Microsoft.Identity.Model, Microsoft.Identity.Web).
+        /// </summary>
+        internal IDictionary<string, string> RawParameters { get; private set; }
+
+        /// <summary>
+        /// Create Authentication-Info parameters from the HttpResponseHeaders for each auth scheme.
+        /// </summary>
+        /// <param name="httpResponseHeaders">HttpResponseHeaders.</param>
+        /// <returns>Authentication-Info provided by the endpoint</returns>
+        public static AuthenticationInfoParameters CreateFromHeaders(HttpResponseHeaders httpResponseHeaders)
+        {
+            AuthenticationInfoParameters parameters = new AuthenticationInfoParameters();
+
+            try
+            {
+                if (httpResponseHeaders.Contains(AuthenticationInfoKey))
+                {
+                    var authInfoValue = httpResponseHeaders.Where(header => header.Key == AuthenticationInfoKey).Single().Value.FirstOrDefault();
```

Comment:
Single() can throw an InvalidOpException if there are 2+ elements. Better to use `Where()` and throw specific exceptions if there is 0 or 2+ elements.

Reply:
I tried to repro this and it is not possible for the HttpResponseHeaders to contain 2 of the same header. It will simply create an array of all of the values assigned to the same key. It still represents one header.